crypto/ring-signature/src/ring_signature/mlsag_sign.rs

```diff
@@ -0,0 +1,343 @@
+#[derive(Debug, Zeroize)]
+pub struct MlsagSignCtx<R: AsMut<[CurveScalar]>> {
+    alpha_0: Scalar,
+    alpha_1: Scalar,
+    G: RistrettoPoint,
+    I: RistrettoPoint,
+
+    /// Cache real input for balance checking
+    #[zeroize(skip)]
+    real_input: Option<(RistrettoPublic, Commitment)>,
+
+    /// Number of rings computed
+    ring_count: usize,
+
+    responses: R,
+
+    real_challenge: Option<Scalar>,
+    last_challenge: Option<Scalar>,
+    zeroth_challenge: Option<Scalar>,
+
+    #[zeroize(skip)]
+    key_image: KeyImage,
+
+    #[zeroize(skip)]
+    output_commitment: Commitment,
+
+    complete: bool,
+}
+
+#[allow(dead_code)]
+impl<R: AsRef<[CurveScalar]> + AsMut<[CurveScalar]>> MlsagSignCtx<R> {
+    /// Initialise signing state, including `challenges` and `responses` working
+    /// buffers
+    pub fn init<'a>(
+        params: &'a MlsagSignParams<'a>,
+        mut rng: impl CryptoRngCore,
+        mut responses: R,
+    ) -> Result<Self, Error> {
+        let G = B_BLINDING;
+        debug_assert!(
+            params.generator.B_blinding == G,
+            "basepoint for blindings mismatch"
+        );
+
+        let r = responses.as_mut();
+
+        // Check buffer lengths are correct for ring size
+        if r.len() != params.ring_size * 2 {
+            return Err(Error::LengthMismatch(params.ring_size * 2, r.len()));
+        }
+
+        // Generate KeyImage from generated transaction private key
+        let key_image = KeyImage::from(params.onetime_private_key);
+
+        // The uncompressed key_image.
+        let I = key_image.point.decompress().ok_or(Error::InvalidKeyImage)?;
+
+        // Uncompressed output commitment.
+        // This ensures that each address and commitment encodes a valid Ristretto
+        // point.
+        let output_commitment =
+            Commitment::new(params.value, *params.output_blinding, params.generator);
+
+        // Responses `r_{0,0}, r_{0,1}, ... , r_{ring_size-1,0}, r_{ring_size-1,1}`.
+        r.iter_mut()
+            .for_each(|v| *v = CurveScalar::from(Scalar::zero()));
+
+        for i in 0..params.ring_size {
+            if i == params.real_index {
+                continue;
+            }
+            r[2 * i].scalar = Scalar::random(&mut rng);
+            r[2 * i + 1].scalar = Scalar::random(&mut rng);
+        }
+
+        Ok(Self {
+            G,
+            I,
+            key_image,
+            output_commitment,
+            alpha_0: Scalar::random(&mut rng),
+            alpha_1: Scalar::random(&mut rng),
+            ring_count: 0,
+            real_input: None,
+            real_challenge: None,
+            last_challenge: None,
+            zeroth_challenge: None,
+            responses,
+            complete: false,
+        })
+    }
+
+    /// Update signing context with provided tx_out.
+    ///
+    /// NOTE THIS _MUST_ FOLLOW A CALL TO INIT AND TRAVERSE THE RING STARTING
+    /// WITH THE REAL TXOUT
```

Comment:
It's annoying and complicated to change/maintain but maybe we could consider leaning on the type system a bit more here? Such that this Cannot Be Called without calling the proper init

Reply:
i though about type-stating the MlsagSignCtx to mitigate that, but because of the need to stream we can't associate any state with the referenced buffers (responses etc.) so there's still a sharp edge there (and you just end up wrapping it in an enum so you can hold it anyway).

could refactor to that, or add an `initialised` flag and runtime check that should get optimised out?